### DISCUSSION TITLE
Vulkan Video is final

### DISCUSSION THREAD

**Discussion by H-G-Hristov** [KhronosGroup/MoltenVK#1812] [Ideas]
https://www.khronos.org/blog/khronos-finalizes-vulkan-video-extensions-for-accelerated-h.264-and-h.265-decode

Are the related extensions on the roadmap? If one is interested to contribute, where should they start (what part of the code)?

VK_KHR_video_queue

**Comment by cdavis5e** [comment]
> Are the related extensions on the roadmap?

They're something we're interested in implementing, particularly since [others have expressed interest][1]. But it's on the back burner right now.

[1]: /KhronosGroup/MoltenVK/issues/1337

> If one is interested to contribute, where should they start (what part of the code)?

To support the Vulkan video extensions, you'd probably need to add an implementation of the `VkVideoSessionKHR` and `VkVideoSessionParametersKHR` objects, which would live in `MoltenVK/MoltenVK/GPUObjects/`. You would also need to extend `MVKPhysicalDevice` to support `vkGetPhysicalDeviceVideoCapabilitiesKHR()`, in `MoltenVK/MoltenVK/GPUObjects/MVKPhysicalDevice.mm`, and any pixel formats that support video coding would need to be indicated in `MoltenVK/MoltenVK/GPUObjects/MVKPixelFormat.mm`. Then you would need to add command objects for any `vkCmd` functions, and those would live in `MoltenVK/MoltenVK/Commands/`. The Vulkan entry points themselves live in `MoltenVK/MoltenVK/Vulkan/vulkan.mm`. Note that you would probably not be using Metal, but the Video Toolbox, which is the lowest level public interface in macOS and iOS for accelerated video decoding and encoding. This means that integrating command execution and synchronizing it with other GPU commands could be problematic.
